src/react/hooks/useReadQuery.ts

```diff
@@ -38,11 +39,25 @@ export interface UseReadQueryResult<TData = unknown> {
 export function useReadQuery<TData>(
   queryRef: QueryReference<TData>
 ): UseReadQueryResult<TData> {
+  const didWarnOnDisposedQueryRef = React.useRef(false);
   const [internalQueryRef, getPromise] = React.useMemo(
     () => unwrapQueryRef(queryRef),
     [queryRef]
   );
 
+  if (!didWarnOnDisposedQueryRef.current && internalQueryRef.disposed) {
+    invariant.warn(
+      `
+'useReadQuery' was called with a disposed queryRef which means the query is no longer watched and cache updates will be missed.
+
+This occurs when calling 'dispose' while 'useReadQuery' is still mounted, either by calling 'dispose' too early, or because you are using React's strict mode and calling 'dispose' in a 'useEffect' cleanup function.
+
+If you're using a queryRef produced by 'useBackgroundQuery' or 'useLoadableQuery', this is a bug in Apollo Client. Please file an issue.
+`
+    );
+    didWarnOnDisposedQueryRef.current = true;
+  }
+
```

Comment:
The `invariant.warn` will only be called (and bundled) if `__DEV__` is set, but the check logic around it will end up in production builds without reason.

(Breaking the rules of hooks based on `__DEV__` here should be fine.)

```suggestion
  const [internalQueryRef, getPromise] = React.useMemo(
    () => unwrapQueryRef(queryRef),
    [queryRef]
  );

  if (__DEV__) {
  const didWarnOnDisposedQueryRef = React.useRef(false);
  if (!didWarnOnDisposedQueryRef.current && internalQueryRef.disposed) {
    invariant.warn(
      `
'useReadQuery' was called with a disposed queryRef which means the query is no longer watched and cache updates will be missed.

This occurs when calling 'dispose' while 'useReadQuery' is still mounted, either by calling 'dispose' too early, or because you are using React's strict mode and calling 'dispose' in a 'useEffect' cleanup function.

If you're using a queryRef produced by 'useBackgroundQuery' or 'useLoadableQuery', this is a bug in Apollo Client. Please file an issue.
`
    );
    didWarnOnDisposedQueryRef.current = true;
  }
}
```

Reply:
Updated in [254e0d4](https://github.com/apollographql/apollo-client/pull/11412/commits/254e0d4bb9d158d53b9363fadfab1f230e48a318). Funny I was going to make this update anyways, but you got to the review before I could push that change 🤣. Thanks for being proactive!